File: mobile-ui-testing-android/app/src/androidTest/java/biz/tearoom6/mobile_ui_testing_android/uitest/helper/UserHelper.java
package biz.tearoom6.mobile_ui_testing_android.uitest.helper;

import android.support.test.espresso.NoMatchingViewException;

import junit.framework.AssertionFailedError;

import biz.tearoom6.mobile_ui_testing_android.R;
import biz.tearoom6.mobile_ui_testing_android.uitest.model.User;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.replaceText;
import static android.support.test.espresso.action.ViewActions.scrollTo;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static biz.tearoom6.mobile_ui_testing_android.uitest.helper.AssetHelper.loadTestUser;
import static biz.tearoom6.mobile_ui_testing_android.uitest.helper.ViewHelper.selectTab;
import static biz.tearoom6.mobile_ui_testing_android.uitest.helper.ViewHelper.waitUntilDisplayed;
import static org.apache.commons.lang3.RandomStringUtils.random;
import static org.apache.commons.lang3.RandomStringUtils.randomAlphabetic;

public class UserHelper {

    public static String temporaryNickname() {
        return "test" + random(3, 'あ', 'ん', false, false) + System.currentTimeMillis();
    }

    public static String temporaryMailAddress() {
        return randomAlphabetic(10).toLowerCase() + System.currentTimeMillis() + "@example.com";
    }

    public static String temporaryPassword() {
        return "password." + randomAlphabetic(4) + System.currentTimeMillis();
    }

    public static void login(String email, String password) {
        selectTab(withText(R.string.tab_user));

        // Input email & password.
        onView(withId(R.id.auth_input_loginEmail)).perform(replaceText((email)));
        onView(withId(R.id.auth_input_loginPassword)).perform(replaceText(password));

        // Click login button.
        onView(withId(R.id.auth_button_login)).perform(scrollTo(), click());
    }

    public static void loginAsRegisteredUser() {
        logout();

        User testUser = loadTestUser();
        login(testUser.getEmail(), testUser.getPassword());

        waitUntilDisplayed(R.id.auth_label_nickname);
    }

    public static void logout() {
        selectTab(withText(R.string.tab_user));
        try {
            onView(withId(R.id.auth_button_logout)).check(matches(isDisplayed())).perform(click());
        } catch (NoMatchingViewException | AssertionFailedError ignore) {
        }
    }

    public static void registerUser(String nickname, String email, String password) {
        selectTab(withText(R.string.tab_user));

        // Input email & password & nickname.
        onView(withId(R.id.auth_input_registerNickname)).perform(replaceText(nickname));
        onView(withId(R.id.auth_input_registerEmail)).perform(replaceText(email));
        onView(withId(R.id.auth_input_registerPassword)).perform(replaceText(password));

        // Click register button.
        onView(withId(R.id.auth_button_register)).perform(scrollTo(), click());
    }
}

File: mobile-ui-testing-appium/spec/support/helpers/user_helper.rb
require 'securerandom'

module UserHelper
  include ElementHelper

  def temporary_nickname
    "test#{SecureRandom.hex(5)}"
  end

  def temporary_email
    "#{SecureRandom.hex(6)}@example.com"
  end

  def temporary_password
    SecureRandom.hex(10)
  end

  def register_user(email: temporary_email, password: temporary_password, nickname: temporary_nickname)
    click_user_tab

    click_button(id: 'auth_button_openRegister')

    # Input email & password & nickname.
    input_text(id: 'auth_input_registerNickname', text: nickname)
    input_text(id: 'auth_input_registerEmail', text: email)
    input_text(id: 'auth_input_registerPassword', text: password)

    # Click register button.
    click_button(id: 'auth_button_register')

    wait_true {
      exists { find_element(*condition(id: 'auth_button_logout')) }
    }
  end

  def login_as_registered_user
    logout
    login(email: user['email'], password: user['password'])
  end

  def login(email:, password:)
    click_user_tab

    # Input email & password.
    input_text(id: 'auth_input_loginEmail', text: email)
    input_text(id: 'auth_input_loginPassword', text: password)

    # Click login button.
    click_button(id: 'auth_button_login')

    wait_true {
      exists { find_element(*condition(id: 'auth_button_logout')) }
    }
  end

  def logout
    click_user_tab
    click_button(id: 'auth_button_logout')

    wait_true {
      exists { find_element(*condition(id: 'auth_button_login')) }
    }
  end

  def error_unauthorized_message
    label(id: 'notes_label_unauthorized')
  end

  def click_user_tab
    click_tab(name: 'User')
  end
end
